tests/ui/type/type-params-in-different-spaces-3.stderr

```diff
@@ -14,6 +14,7 @@ LL |         u
               found type parameter `X`
    = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
    = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
+   = note: the caller chooses a type for `Self` which can be different from `X`
```

Comment:
Ah, this one is interesting. The note is technically speaking correct but I don't know if it helps the user. Remember that we add the note “the caller chooses …” because the user might not understand that `fn f<T>() -> T` doesn't mean “can return anything the call*ee* / the author of the function wants to return”. What do you think? Do you think it's useful?

Reply:
This note doesn't seem very helpful in this instance, no. This particular case probably needs a separate note suggesting that `X` might not satisfy `X: Sized` because `Self: Sized` or something to that effect. Not entirely sure what help to give here though.

Reply:
Let's not block this PR on this (given the long history of this diagnostic change with several failed PRs), we should probably not emit the note if the param is `kw::SelfUpper` but that can be done in a follow up.

Sorry for the delay, I was sitting at my desk wondering why we're not suggesting impl-Trait on `fn f<T>() -> (T,) { (0,) }` etc. (so, `fn f() -> (impl Sized,) { (0,) }` which works perfectly). I should probably open an issue for that. If we end up making `try_suggest_return_impl_trait` smarter at some point we can then probably also move the note back into impl-Trait since it does some other checks, too, which we don't do for `note_caller_chooses_ty_for_ty_param`.

> suggesting that X might not satisfy X: Sized but Self: Sized or something to that effect

Nah, that's not relevant here. The `Sized` bounds only exist to allow the method to return `self`.